### PR TITLE
[High priority][WIP] Added BF16 functionality inside paddle core

### PR DESCRIPTION
**Pull request by jakpiase**
### PR types
Others

### PR changes
Others

### Describe
Added BF16 functionality inside paddle core. After merging this PR uint16 won't be used as a placeholder for BF16 datatype anymore, so issues: #39473 and #34927 should be fixed

### CODE REVIEW

File: python/paddle/fluid/framework.py [hunk 1]
```diff
@@ -992,7 +993,7 @@ def convert_np_dtype_to_dtype_(np_dtype):
         return core.VarDesc.VarType.INT64
     elif dtype == np.bool:
         return core.VarDesc.VarType.BOOL
-    elif dtype == np.uint16:
+    elif dtype == bfloat16:
         # since there is still no support for bfloat16 in NumPy,
         # uint16 is used for casting bfloat16
         return core.VarDesc.VarType.BF16
```
Comment (lidanqing-vv) [inline python/paddle/fluid/framework.py:1179]: Hi, Do you think if it is necessary to map np.uint16 to core.VarDesc.VarType.UINT16?  Or maybe make it reserved datatype? Also please modify the comments ?
I think I saw more places that used uint16 in python than now, not in tests

Comment (jakpiase) [reply python/paddle/fluid/framework.py:1179]: Hi, this PR is not ready yet, I have made it just to see which tests and CIs will fail

Comment (lidanqing-vv) [reply python/paddle/fluid/framework.py:1179]: Hi, ok, thanks

**Comment by lidanqing-vv** [comment]
Hi, @jakpiase there is file conflict, could you please rebase and solve the conflict? Thank you. 
After this PR is merged, please also cherry-pick it into release/2.3, this PR can catch up 2.3 formal release. Thank you very much for your work !

File: cmake/flags.cmake [hunk 2]
```diff
@@ -133,7 +133,6 @@ if (NOT WIN32)
 set(COMMON_FLAGS
     -fPIC
     -fno-omit-frame-pointer
-    -Werror
```
Comment (jczaja) [inline cmake/flags.cmake]: is this removing of -Werror intentional?

Comment (jakpiase) [reply cmake/flags.cmake]: It's only temporary to silence paddle_bfloat compilation warnings, that won't be included into the final version of this PR

**Comment by Aganlengzi** [comment]
@jakpiase hi, thanks for you work, you can attach https://github.com/PaddlePaddle/Paddle/issues/39473 and https://github.com/PaddlePaddle/Paddle/issues/34927 in Describe of this PR to synchronize states automatically.

**Comment by jakpiase** [comment]
The work on that functionality was cancelled due to some issues on Windows-GPU evironment. Unfortunately I was not able to debug that and because of that I have aborted my work, but except for that case my current changes(after resurrecting this PR to the state of modern PaddlePaddle code) should enable numpy's bfloat16 datatype inside paddle core and enable passing that type of tensors both between python-c++ and c++-python.

**Pull request closed without merging**